File: lab/adversary/mutation_test.py
#!/usr/bin/env python3
"""Mutation test for the adversary harness (docs/GATE1.md).

Runs the toy system once correctly and once per deliberate defect. For each run:
the server container captures its own traffic with tcpdump while storing
messages in SQLite and logging headers; the client sends canary-bearing
messages, fetches them, keeps and purges a local history. Then canary_scan.py
scans everything an adversary could hold: server DB, server log, packet capture,
and the client's purged local DB.

Pass: the correct run is CLEAN (no false positives) and every defect is CAUGHT
in the artefact where it should appear. Results: results/<label>.jsonl, flushed.
Usage: mutation_test.py LABEL
"""
import json, os, secrets, shutil, subprocess, sys, time
HERE = os.path.dirname(os.path.abspath(__file__))
IMG, SRV, CLI = "tremulator-toy:latest", "tadv-srv", "tadv-cli"
MODES = {  # defect: artefacts where it must appear (any one suffices)
    "none": [],
    "plaintext_body": ["traffic.pcap", "server.db"],
    "plaintext_attachment": ["traffic.pcap", "server.db"],
    "preview_leak": ["traffic.pcap", "server.db"],
    "purge_unlink_only": ["client.db"],
    "key_in_header": ["server.log", "traffic.pcap"],
    "base64_only": ["traffic.pcap", "server.db"],
    "gzip_only": ["traffic.pcap", "server.db"],
}
label = sys.argv[1]
os.makedirs(os.path.join(HERE, "results"), exist_ok=True)
res = open(os.path.join(HERE, "results", f"{label}.jsonl"), "a")
def log(r):
    r["t"] = time.strftime("%Y-%m-%dT%H:%M:%S"); res.write(json.dumps(r) + "\n"); res.flush(); print(json.dumps(r), flush=True)
def sh(cmd, timeout=180):
    return subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)

# One network and one container per role, created ONCE and reused across modes;
# each mode starts a fresh server process and capture inside the same container.
# The first version made a network and two containers per mode, and that
# interface churn raised "disconnected" desktop alerts on Grey's laptop
# (2026-09-21).
NET = "tadv-net"
outroot = os.path.join(HERE, "out", label)
os.makedirs(outroot, exist_ok=True)
def teardown():
    sh(["docker", "rm", "-f", SRV, CLI]); sh(["docker", "network", "rm", NET])
teardown()
sh(["docker", "network", "create", NET])
for c in (SRV, CLI):
    r = sh(["docker", "run", "-d", "--name", c, "--network", NET, "-v", f"{outroot}:/out", IMG, "sleep", "infinity"])
    if r.returncode: raise RuntimeError(r.stderr)
try:
    for mode, expect in MODES.items():
        base = os.path.join(outroot, mode); shutil.rmtree(base, ignore_errors=True)
        inp, srvd, clid = (os.path.join(base, d) for d in ("in", "srv", "cli"))
        for d in (inp, srvd, clid): os.makedirs(d)
        cb = f"/out/{mode}"
        key = secrets.token_hex(32)
        msgs = [f"m{i}\tCANARY-{secrets.token_hex(8)}" for i in range(5)]
        open(os.path.join(inp, "msgs.tsv"), "w").write("\n".join(msgs) + "\n")
        open(os.path.join(base, "canaries.tsv"), "w").write("\n".join(msgs + [f"key\t{key}"]) + "\n")
        sh(["docker", "exec", "-d", SRV, "sh", "-c",
            f"tcpdump -i eth0 -U -w {cb}/srv/traffic.pcap & sleep 1; "
            f"TOY_DB={cb}/srv/server.db TOY_LOG={cb}/srv/server.log exec python3 /toy/server.py"])
        time.sleep(3)
        r = sh(["docker", "exec", "-e", f"DEFECT={mode}", "-e", f"TOY_KEY_HEX={key}",
                "-e", f"TOY_LOCAL_DB={cb}/cli/client.db", CLI, "python3", "/toy/client.py", SRV, f"{cb}/in/msgs.tsv"])
        client_out = (r.stdout + r.stderr).strip()[-300:]
        time.sleep(1)
        sh(["docker", "exec", SRV, "sh", "-c", "pkill -INT tcpdump; pkill -f /toy/server.py; sleep 1"])
        hits_path = os.path.join(base, "hits.jsonl")
        sh(["python3", os.path.join(HERE, "canary_scan.py"), os.path.join(base, "canaries.tsv"), hits_path, srvd, clid])
        hits = [json.loads(l) for l in open(hits_path)] if os.path.exists(hits_path) else []
        where = sorted({os.path.basename(h["where"].split("#")[0].split("|")[0]) for h in hits})
        artefacts = sorted(os.listdir(srvd) + os.listdir(clid))
        ok = (not hits) if mode == "none" else any(a in where for a in expect)
        log({"run": label, "mode": mode, "client": client_out, "artefacts": artefacts,
             "hits": len(hits), "hit_in": where, "expected_in": expect,
             "verdict": "CLEAN" if not hits else "CAUGHT", "pass": ok})
finally:
    teardown()
res.close()
# Containers write as root; hand the outputs back to the invoking user.
sh(["docker", "run", "--rm", "-v", f"{os.path.join(HERE, 'out')}:/o", "tremulator-lab:latest",
    "chown", "-R", f"{os.getuid()}:{os.getgid()}", "/o"])
print("DONE", flush=True)

File: lab/tool/bin/mutation_test.dart
/// Mutation test for the adversary harness (docs/GATE1.md), ported from
/// lab/adversary/mutation_test.py.
///
/// Runs the Dart toy once correctly and once per deliberate defect. For each
/// mode the server container captures its own traffic with tcpdump while
/// storing messages in SQLite and logging headers; the client sends
/// canary-bearing messages, fetches them, keeps and purges a local history.
/// Then the canary scanner reads everything an adversary could hold: server
/// DB, server log, packet capture, and the client's purged local DB.
///
/// Pass: the correct run is CLEAN and every defect is CAUGHT where it should
/// appear. One network and one container per role are created once and reused,
/// because per-mode create/remove raised "disconnected" alerts on Grey's
/// laptop. Usage: dart run bin/mutation_test.dart LABEL
library;

import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'package:tremulator_lab/canary_scan.dart';

const img = 'tremulator-toy-dart:latest';
const srv = 'tadv-srv';
const cli = 'tadv-cli';
const net = 'tadv-net';
const modes = <String, List<String>>{
  'none': [],
  'plaintext_body': ['traffic.pcap', 'server.db'],
  'plaintext_attachment': ['traffic.pcap', 'server.db'],
  'preview_leak': ['traffic.pcap', 'server.db'],
  'purge_unlink_only': ['client.db'],
  'key_in_header': ['server.log', 'traffic.pcap'],
  'base64_only': ['traffic.pcap', 'server.db'],
  'gzip_only': ['traffic.pcap', 'server.db'],
};

final _rnd = Random.secure();
String _hex(int n) => List.generate(
  n,
  (_) => _rnd.nextInt(256).toRadixString(16).padLeft(2, '0'),
).join();

Future<ProcessResult> sh(List<String> c) => Process.run(c.first, c.sublist(1));

Future<void> teardown() async {
  await sh(['docker', 'rm', '-f', srv, cli]);
  await sh(['docker', 'network', 'rm', net]);
}

Future<void> main(List<String> args) async {
  final label = args.first;
  final lab = File(Platform.script.toFilePath()).parent.parent.parent.path;
  final outRoot = Directory('$lab/adversary/out/$label')
    ..createSync(recursive: true);
  final results = File(
    '$lab/adversary/results/$label.jsonl',
  ).openWrite(mode: FileMode.append);
  await teardown();
  await sh(['docker', 'network', 'create', net]);
  for (final c in [srv, cli]) {
    final r = await sh([
      'docker',
      'run',
      '-d',
      '--name',
      c,
      '--network',
      net,
      '-v',
      '${outRoot.path}:/out',
      img,
      'sleep',
      'infinity',
    ]);
    if (r.exitCode != 0) throw StateError('${r.stderr}');
  }
  try {
    for (final e in modes.entries) {
      final mode = e.key;
      final base = Directory('${outRoot.path}/$mode');
      if (base.existsSync()) base.deleteSync(recursive: true);
      for (final d in ['in', 'srv', 'cli']) {
        Directory('${base.path}/$d').createSync(recursive: true);
      }
      final cb = '/out/$mode';
      final key = _hex(32);
      final msgs = [for (var i = 0; i < 5; i++) 'm$i\tCANARY-${_hex(8)}'];
      File(
        '${base.path}/in/msgs.tsv',
      ).writeAsStringSync('${msgs.join('\n')}\n');
      final canaries = '${base.path}/canaries.tsv';
      File(
        canaries,
      ).writeAsStringSync('${[...msgs, 'key\t$key'].join('\n')}\n');
      final serverCmd =
          'tcpdump -i eth0 -U -w $cb/srv/traffic.pcap & sleep 1; '
          'TOY_DB=$cb/srv/server.db TOY_LOG=$cb/srv/server.log '
          'exec /toy/server/bin/toy_server';
      await sh(['docker', 'exec', '-d', srv, 'sh', '-c', serverCmd]);
      await Future<void>.delayed(const Duration(seconds: 3));
      final r = await sh([
        'docker',
        'exec',
        '-e',
        'DEFECT=$mode',
        '-e',
        'TOY_KEY_HEX=$key',
        '-e',
        'TOY_LOCAL_DB=$cb/cli/client.db',
        cli,
        '/toy/client/bin/toy_client',
        srv,
        '$cb/in/msgs.tsv',
      ]);
      final clientOut = '${r.stdout}${r.stderr}'.trim();
      await Future<void>.delayed(const Duration(seconds: 1));
      await sh([
        'docker',
        'exec',
        srv,
        'sh',
        '-c',
        'pkill -INT tcpdump; pkill -f toy_server; sleep 1',
      ]);
      // hand the outputs back to the invoking user before scanning
      await sh([
        'docker',
        'exec',
        srv,
        'chown',
        '-R',
        '${await _id('-u')}:${await _id('-g')}',
        cb,
      ]);
      final hitsSink = File('${base.path}/hits.jsonl').openWrite();
      final hits = await scanPaths(canaries, [
        '${base.path}/srv',
        '${base.path}/cli',
      ], hitsSink);
      await hitsSink.close();
      final where =
          hits
              .map(
                (h) =>
                    h.where.split('#').first.split('|').first.split('/').last,
              )
              .toSet()
              .toList()
            ..sort();
      final artefacts = [
        for (final d in ['srv', 'cli'])
          ...Directory(
            '${base.path}/$d',
          ).listSync().map((f) => f.uri.pathSegments.last),
      ]..sort();
      final ok = mode == 'none' ? hits.isEmpty : e.value.any(where.contains);
      final rec = {
        'run': label,
        'mode': mode,
        'client': clientOut.substring(max(0, clientOut.length - 300)),
        'artefacts': artefacts,
        'hits': hits.length,
        'hit_in': where,
        'expected_in': e.value,
        'verdict': hits.isEmpty ? 'CLEAN' : 'CAUGHT',
        'pass': ok,
        't': DateTime.now().toIso8601String().substring(0, 19),
      };
      results.writeln(jsonEncode(rec));
      await results.flush();
      print(jsonEncode(rec));
    }
  } finally {
    await teardown();
    await results.close();
  }
}

Future<String> _id(String flag) async =>
    ((await Process.run('id', [flag])).stdout as String).trim();
